### PR TITLE
Plot widget improvements

### PR DESCRIPTION
**Pull request by mattgibbs**
This pull request adds support for multiple curves in PyDMTimePlot and PyDMWaveformPlot.  In designer, you can add or remove curves from a plot by right clicking it and choosing 'Edit Curves', or by double-clicking the plot.  This will bring up a dialog that lets you edit the curves.  All curves info is stored in a QStringList property called 'curves', each item in the list is a JSON dictionary with properties for the curve.  You can edit this directly, if you like, but using the Edit Curves interface is nicer.

Manipulating curves programmatically is also possible - each plot widget has methods for adding and removing curves in code.

This is a pretty large change for plots - people using PyDM in the wild (@laispc, @gabrielfedel, @mennodoor) may want to try pulling this branch and testing it out.

### CODE REVIEW

File: pydm/widgets/waveformplot.py [hunk 1]
```diff
@@ -5,85 +5,204 @@
...
+	def __init__(self, parent=None, init_x_channels=[], init_y_channels=[], background='default'):
+		super(PyDMWaveformPlot, self).__init__(parent, background)
+		#If the user supplies a single string instead of a list, wrap it in a list.
+		if isinstance(init_x_channels, str):
+			init_x_channels = [init_x_channels]
+		if isinstance(init_y_channels, str):
+			init_y_channels = [init_y_channels]
+		if len(init_x_channels) == 0:
+			init_x_channels = list(itertools.repeat(None, len(init_y_channels)))
+		if len(init_x_channels) > 0 and len(init_x_channels) != len(init_y_channels):
```
Comment (hhslepicka) [inline pydm/widgets/waveformplot.py]: you can probably remove the first check: ```len(init_x_channels) > 0``` can't you?

File: pydm/widgets/waveformplot_table_model.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
+from PyQt4.QtCore import QAbstractTableModel, Qt, QModelIndex, QVariant
```
Comment (hhslepicka) [inline pydm/widgets/waveformplot_table_model.py]: Forced to PyQt4... change to PyDM PyQt import to solve PyQT4, 5, etc...

File: pydm/widgets/waveformplot_table_model.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+from PyQt4.QtGui import QBrush, QColor
```
Comment (hhslepicka) [inline pydm/widgets/waveformplot_table_model.py]: Same as above

File: pydm/widgets/timeplot_table_model.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
+from PyQt4.QtCore import QAbstractTableModel, Qt, QModelIndex, QVariant
```
Comment (hhslepicka) [inline pydm/widgets/timeplot_table_model.py]: Use pydm PyQT import and not PyQt4...

File: pydm/widgets/timeplot_table_model.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+from PyQt4.QtGui import QBrush, QColor
```
Comment (hhslepicka) [inline pydm/widgets/timeplot_table_model.py]: Same as above

File: pydm/widgets/timeplot.py [hunk 6]
```diff
@@ -1,160 +1,255 @@
...
+    if self.points_accumulated < self._bufferSize:
+      self.points_accumulated = self.points_accumulated + 1    
+    
+  def initialize_buffer(self):
+    self.points_accumulated = 0
+    #If you don't specify dtype=float, you don't have enough resolution for the timestamp data.
+    self.data_buffer = np.zeros((2,self._bufferSize), order='f',dtype=float)
+    self.data_buffer[0].fill(time.time())
+    
+  def getBufferSize(self):
```
Comment (laispc) [inline pydm/widgets/timeplot.py]: @hhslepicka, a pyqtproperty for the attribute bufferSize is already in PyDMTimePlot class. :)
I guess that TimePlotCurveItem class does not need pyqtproperty declarations since its attributes are not shown in Designer.

File: pydm/widgets/baseplot.py [hunk 7]
```diff
@@ -101,14 +136,37 @@ def setAxisColor(self, color):
   axisColor = pyqtProperty(QColor, getAxisColor, setAxisColor)
   
   def getPlotTitle(self):
+    if self._title is None:
+      return ""
     return str(self._title)
   
   def setPlotTitle(self, value):
     self._title = str(value)
+    if len(self._title) < 1:
```
Comment (mattgibbs) [inline pydm/widgets/baseplot.py]: Initially my thought was that comparing to an empty string literal was maybe slightly slower, but timeit benchmarks prove me wrong.  Not that it matters for something that gets called quite rarely anyway.

File: pydm/widgets/timeplot.py [hunk 8]
```diff
@@ -1,160 +1,255 @@
...
+      self._update_mode = PyDMTimePlot.AsynchronousMode
+    else:
+      self._update_mode = PyDMTimePlot.SynchronousMode
+    self.initialize_buffer()
+
+  def resetUpdatesAsynchronously(self):
+    self._update_mode = PyDMTimePlot.SynchronousMode
+    self.initialize_buffer()
+  
+  def max_x(self):
```
Comment (laispc) [inline pydm/widgets/timeplot.py]: I believe that the idea of this method it to return the maximum value of x dimension (time values). It is returning a float. max_x() is used in updateXAxis to find the maximum time value among all curves, so the plot can be adjusted to fit all curves inside.

File: pydm/widgets/timeplot.py [hunk 9]
```diff
@@ -1,160 +1,255 @@
...
+  
+  @address.setter
+  def address(self, new_address):
+    if new_address is None or len(str(new_address)) < 1:
+      self.channel = None
+      return
+    self.channel = PyDMChannel(address=new_address, connection_slot=self.connectionStateChanged, value_slot=self.receiveNewValue)
+  
+  @property
+  def color_string(self):
```
Comment (mattgibbs) [inline pydm/widgets/timeplot.py]: This one is just a python property, and isn't exposed as a PyQt property.

File: pydm/widgets/timeplot.py (hunk 6 above)
Comment (mattgibbs) [inline pydm/widgets/timeplot.py]: @laispc is right, the TimePlotCurveItem is supposed to be internal to the PyDMTimePlot class, not really something that gets exposed to users.

File: pydm/widgets/timeplot.py [hunk 10]
```diff
@@ -1,160 +1,255 @@
...
+    if self._bufferSize != 1200:
+      self._bufferSize = 1200
+      self.initialize_buffer()
+  
+  @pyqtSlot()
+  def redrawCurve(self):
+    if self.connected:
+      self.setData(y=self.data_buffer[1, -self.points_accumulated:], x=self.data_buffer[0, -self.points_accumulated:])
+  
+  def setUpdatesAsynchronously(self, value):
```
Comment (mattgibbs) [inline pydm/widgets/timeplot.py]: Ditto above.  The PyDMTimePlot has a PyQt property for this, and it just forwards that to all the curves it holds.  (This is supposed to keep the potential for the option to mix synchronous and asynchronous updating curves if we need it in the future).

File: pydm/widgets/timeplot_curve_editor.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
+from ..PyQt.QtGui import QDialog, QVBoxLayout, QHBoxLayout, QTableView, QAbstractItemView, QSpacerItem, QSizePolicy, QDialogButtonBox, QPushButton, QItemSelection
+from ..PyQt.QtCore import Qt, pyqtSlot
```
Comment (laispc) [inline pydm/widgets/timeplot_curve_editor.py]: For me, it only works when `QItemSelection` is imported from `..PyQt.QtCore`. I mean:

`..PyQt.QtCore import Qt, pyqtSlot, QItemSelection`

File: pydm/widgets/timeplot.py [hunk 12]
```diff
@@ -1,160 +1,255 @@
...
+    if value == True:
```
Comment (mattgibbs) [inline pydm/widgets/timeplot.py]: This is an interesting one to me:  in cases where I am expecting a boolean true or false, I like to check if it is actually True, and not just a 'true-ish' value.  In this example, I don't think it makes much of a difference, but I think it has the potential to save you from strange issues where you are passing in the completely wrong time into a function that is supposed to accept a boolean.

File: pydm/widgets/timeplot_qtplugin.py [hunk 13]
```diff
@@ -1,5 +1,75 @@
-from .qtplugin_base import qtplugin_factory
+from ..PyQt.QtDesigner import QExtensionFactory, QPyDesignerTaskMenuExtension
+from ..PyQt.QtGui import QAction
+from ..PyQt.QtCore import pyqtSlot
+from .qtplugin_base import PyDMDesignerPlugin
 from .timeplot import PyDMTimePlot
+from timeplot_curve_editor import TimePlotCurveEditorDialog
```
Comment (laispc) [inline pydm/widgets/timeplot_qtplugin.py]: I only got this line working by relative import (adding the dot):

`from .timeplot_curve_editor import TimePlotCurveEditorDialog`

File: pydm/widgets/timeplot.py (hunk 8 above)
Comment (mattgibbs) [inline pydm/widgets/timeplot.py]: In this case, data_buffer[0] is the time axis values, and is always monotonically increasing.  Getting data_buffer[0, -1] gives you the last point of the time axis, which is always the max value.

File: pydm/widgets/timeplot_curve_editor.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+from ..PyQt.QtDesigner import QDesignerFormWindowInterface
+from timeplot_table_model import PyDMTimePlotCurvesModel
```
Comment (laispc) [inline pydm/widgets/timeplot_curve_editor.py]: I only got this line working by relative import (adding the dot):

`from timeplot_table_model import PyDMTimePlotCurvesModel`

File: pydm/widgets/timeplot_table_model.py [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+from operator import itemgetter
+
+from timeplot import TimePlotCurveItem
```
Comment (laispc) [inline pydm/widgets/timeplot_table_model.py]: Again, I would add a relative import here:

`from .timeplot import TimePlotCurveItem`

File: pydm/widgets/waveformplot_curve_editor.py [hunk 16]
```diff
@@ -0,0 +1,72 @@
+from ..PyQt.QtGui import QDialog, QVBoxLayout, QHBoxLayout, QTableView, QAbstractItemView, QSpacerItem, QSizePolicy, QDialogButtonBox, QPushButton, QItemSelection
+from ..PyQt.QtCore import Qt, pyqtSlot
```
Comment (laispc) [inline pydm/widgets/waveformplot_curve_editor.py]: I would import QItemSelection from QtCore here also:

`from ..PyQt.QtCore import Qt, pyqtSlot, QItemSelection`

File: pydm/widgets/waveformplot_curve_editor.py [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+from ..PyQt.QtDesigner import QDesignerFormWindowInterface
+from waveformplot_table_model import PyDMWaveformPlotCurvesModel
```
Comment (laispc) [inline pydm/widgets/waveformplot_curve_editor.py]: I would add a relative import here also:
`from .waveformplot_table_model import PyDMWaveformPlotCurvesModel`

File: pydm/widgets/baseplot.py (hunk 7 above)
Comment (hhslepicka) [inline pydm/widgets/baseplot.py]: No problem at all! Both look good to me! 👍

File: pydm/widgets/timeplot.py (hunk 9 above)
Comment (hhslepicka) [inline pydm/widgets/timeplot.py]: Oh... my bad! Eyes got tricked on this one! I got used to type the pyqtproperty that I see property and automatically relate to pyqtproperty!

File: pydm/widgets/timeplot.py (hunk 6 above)
Comment (hhslepicka) [inline pydm/widgets/timeplot.py]: That is true! Just ignore this comment! Thank you @laispc and @mattgibbs

File: pydm/widgets/timeplot_table_model.py [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+		if not index.isValid():
+			return False
+		if index.row() >= self.rowCount():
+			return False
+		if index.column() >= self.columnCount():
+			return False
+		column_name = self._column_names[index.column()]
+		curve = self.plot._curves[index.row()]
+		if column_name == "Channel":
+			curve.address = str(value.toString())
```
Comment (laispc) [inline pydm/widgets/timeplot_table_model.py]: Here, this line lead to error: `AttributeError: type object 'str' has no attribute 'toString'`
Changing to 
`curve.address = str(value)`
it works fine.

File: pydm/widgets/timeplot.py (hunk 12 above)
Comment (hhslepicka) [inline pydm/widgets/timeplot.py]: I agree! 👍

File: pydm/widgets/timeplot_table_model.py [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+		elif column_name == "Label":
+			curve.setData(name=str(value.toString()))
```
Comment (laispc) [inline pydm/widgets/timeplot_table_model.py]: Same as above.

File: pydm/widgets/timeplot_table_model.py [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+		elif column_name == "Color":
+			curve.color_string = str(value.toString())
```
Comment (laispc) [inline pydm/widgets/timeplot_table_model.py]: Same as above.

File: pydm/widgets/timeplot.py (hunk 8 above)
Comment (hhslepicka) [inline pydm/widgets/timeplot.py]: Again my eyes tricking me... I saw this ```[0, -1]``` and my brain interpreted this: ```[0:-1]```. I need glasses for sure.

File: pydm/utilities/colors.py [hunk 21]
```diff
@@ -0,0 +1,25 @@
+import pickle
+import os
+import itertools
+
+current_dir = os.path.dirname(os.path.realpath(__file__))
+
+svg_color_to_hex_map = None
+hex_to_svg_color_map = None
+with open(os.path.join(current_dir, 'hex2color.pkl')) as f:
+  hex_to_svg_color_map = pickle.load(f)
```
Comment (laispc) [inline pydm/utilities/colors.py]: This line lead to error:  `TypeError: a bytes-like object is required, not 'str'`

Opening the file in binary mode solved it. I mean:
`with open(os.path.join(current_dir, 'hex2color.pkl'), 'rb') as f:`

File: pydm/utilities/colors.py [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+with open(os.path.join(current_dir, 'color2hex.pkl')) as f:
+  svg_color_to_hex_map = pickle.load(f)
```
Comment (laispc) [inline pydm/utilities/colors.py]: Same as above.

File: pydm/widgets/timeplot_curve_editor.py (hunk 11 above)
Comment (hhslepicka) [inline pydm/widgets/timeplot_curve_editor.py]: @laispc which version of PyQT are you using? 4.8? 4.4? 5?
For PyQt4... QItemSelection is part of QtGui: http://pyqt.sourceforge.net/Docs/PyQt4/qitemselection.html

And for PyQt5 it is part of QtCore...
https://doc.qt.io/qt-5/qitemselection.html

@mattgibbs ... what about add a variable on the PyQt module inside PyDM to define the version of PyQT being used and do this check?

File: pydm/widgets/waveformplot_curve_editor.py (hunk 16 above)
Comment (hhslepicka) [inline pydm/widgets/waveformplot_curve_editor.py]: Same issue as above

File: pydm/widgets/timeplot_table_model.py (hunk 4 above)
Comment (mattgibbs) [inline pydm/widgets/timeplot_table_model.py]: Nice find!  I just fixed it and pushed.

**Comment by laispc** [comment]
@mattgibbs, the changes you implemented are awesome. The "Edit Curves" menu in Designer is very friendly. I am adding some review comments. Thanks.

File: pydm/widgets/timeplot_curve_editor.py (hunk 11 above)
Comment (mattgibbs) [inline pydm/widgets/timeplot_curve_editor.py]: Luckily, the variable exists.  I'm going to make it so that you can import QItemSelection from PyQt.QtCore OR PyQt.QtGui, so that people used to PyQt4 and PyQt5 both get the behavior the expect.

Comment (laispc) [inline pydm/widgets/timeplot_curve_editor.py]: @hhslepicka I am currently using PyQt 5.7 (with Python 3.5). You are right, importing QItemSelection differs in PyQt4 and PyQt5. I agree that a version checking would be helpful here. :)

File: pydm/widgets/waveformplot.py [hunk 23]
```diff
@@ -5,85 +5,204 @@
...
+	data_changed = pyqtSignal()
+	def __init__(self, y_addr=None, x_addr=None, **kws):
+		if y_addr is None:
+			y_addr = ""
+		if 'name' not in kws or kws['name'] is None:
+			try:
+				y_name = y_addr.split("://")[1]
+			except IndexError:
+				y_name = y_addr
+			if x_addr is None:
```
Comment (mattgibbs) [inline pydm/widgets/waveformplot.py]: It is actually correct, but it is very hard to follow.  This is pretty ugly, but its trying to do this:  If a name wasn't specified at all, try to use the y_addr and/or x_addr to make a default name.  If there is only a y_addr provided, use just that.  If both y_addr and x_addr were provided, use them both.

File: pydm/widgets/waveformplot.py (hunk 1 above)
Comment (mattgibbs) [inline pydm/widgets/waveformplot.py]: Yep!  That check was left over from an old version.  It is always satisfied.  I'll remove it.

File: pydm/widgets/waveformplot.py (hunk 23 above)
Comment (hhslepicka) [inline pydm/widgets/waveformplot.py]: Let me know what you think about this:
```python
 def __init__(self, y_addr=None, x_addr=None, **kws):
     def remove_protocol(addr):
         try:
             name = addr.split("://")[1]
         except IndexError:
             name = addr
         return name

     y_addr = "" if y_addr is None else y_addr

     if kws.get('name') is None:
         y_name = remove_protocol(y_addr)
         if x_addr is None:
             plot_name = y_name
         else:
             x_name = remove_protocol(x_addr)
             plot_name = "{y} vs. {x}".format(y=y_name, x=x_name)
         kws['name'] = plot_name

     self.x_channel = None
     self.y_channel = None
     self.x_address = x_addr
     self.y_address = y_addr
     self.x_waveform = None
     self.y_waveform = None
     super(WaveformCurveItem, self).__init__(**kws)
```

Comment (hhslepicka) [inline pydm/widgets/waveformplot.py]: Also, it would be very good to have this ```remove_protocol``` in the utils of PyDM since it will be used in more places! ;)

**Comment by mattgibbs** [comment]
Hi @laispc and @hhslepicka, I just pushed some changes that I think fix all the remaining crashing problems.  Mind giving it another look?

**Comment by hhslepicka** [comment]
@mattgibbs all looks pretty good to me! Feel free to address or not the comment in this part: https://github.com/slaclab/pydm/pull/60#discussion_r128049357

File: pydm/widgets/waveformplot.py (hunk 23 above)
Comment (mattgibbs) [inline pydm/widgets/waveformplot.py]: Done!

File: pydm/widgets/timeplot_table_model.py [hunk 24]
```diff
@@ -0,0 +1,106 @@
+from ..PyQt.QtCore import QAbstractTableModel, Qt, QModelIndex, QVariant
+from ..PyQt.QtGui import QBrush, QColor
+from operator import itemgetter
+
+from timeplot import TimePlotCurveItem
```
Comment (laispc) [inline pydm/widgets/timeplot_table_model.py]: Missing relative import here:
`from .timeplot import TimePlotCurveItem`

**Comment by laispc** [comment]
Hi @mattgibbs and @hhslepicka. The only thing missing is a relative import (see comment above). Besides that, everything working fine here! Thanks. :)

**Comment by hhslepicka** [comment]
1e7fe5a220b4cc1ddf53124c19553a0d--happy-birthday-meme-birthday-memes
